Not everyone wants to enter dates with the day first

lenlo opened the pull request:
My preferred date format follows the ISO standard (YYYY-MM-DD), but moneyGuru insists on me entering them with the day first, then month, then year, i.e. backwards from how they are written and how I normally would enter them. At first I thought this was a simple bug and fixed it to always be left-to-right in whatever format the user prefers, but then I came across an article on the moneyGuru website that promoted the day->month->year entry as a feature, so I added a preference that allowed the old behavior to be reenabled. I still let the default be left-to-right as this would arguably be the most "natural" and expected behavior.

hsoft commented:
I don't have enough time to review this one right now, but it looks interesting! Will probably be useful to many. My preliminary comments are:

* I'm lukewarm about having this preference being stored as a class variable. I know that because of the way that the `DateWidget` is instantiated, it might be tough to do otherwise, but still, lukewarm :)
* It seems over-engineered to me to hold a whole *date format* for date entry. A simple `bool` would do I think because nobody's ever going to want anything else than the display format as their entry format (except for the original `d -> m -> y` of course :) )!
* Unit tests.

lenlo commented:
I'm lukewarm about the whole idea to do date entry in any other way than left-to-right, but sure... ;-)

Storing the override in the class seems like the obvious place to me, but if you have a better suggestion I'm open for it. I assume you don't want the class to reach out to the app and grab it from there. That would introduce an unnecessary dependency and make the code less compartmentalized.
 
I agree that the entry_format parameter to the DateWidget's constructor might be overkill and I could remove that if you want (or feel free to do so yourself). But that won't actually make much of a difference to the rest of the code. As long as DateWidgets take a format parameter, they also need to remember which order to fill in the day/month/year fields on an individual basis.

The unit tests. OK. Let me look into that...

lenlo commented:
Updated the DateWidget tests to cover both date entry orders d->m->y and left-to-right (for yyyy-mm-dd).

hsoft commented:
I'm not sure we're talking about the same thing with regards to my 2nd point. What I meant is not to remove the constructor, but rather to make it a `bool` rather than a format. It would make the mechanism much simpler. Everything would stay pretty much the same except that if the flag is `True`, a `_next()` call would become the equivalent of calling `right()`.

Also, it seems that tests you've added are false positive. `left()` and `right()` already work independently if the entry format. Those tests are redundant with others because they weren't failing before you added your changes. A more meaningful test would be:

```
    def test_type_2(self):
        self.w.type('2')
        eq_(self.w.text, '2   -06-12')
```

(and it would help to add the same test in ` TestCaseYYYYMMDDWithDot` because it seems that we lacked that coverage)

lenlo commented:
Yeah, you're right, `left_left()` and `test_right()` are arguably redundant (although `test_selection()` is still useful).

I've now replaced `left` and `right` with a `type` test as per your suggestions (albeit with a few more characters) and added a similar one to `TestCaseYYYYMMDDWithDot`.

I'm sorry, but I'm less convinced about the value of replacing the `entry_format` parameter with a `bool`. You can't connect it to the preference as there may already be a plethora of DateWidget objects that already have been instantiated on the screen (and won't change) when/if the preference is changed. This means there won't be any change of behavior for them until the sheet is closed and reopened. However, you *could* dump the `entry_format` constructor parameter and change the `setEntryFormat` class method to (say) `setDMYEntry` which takes a `bool` instead of a `format`. I'm a bit hesitant as the code for the more general case already has been written and I would argue that it's not *that* complicated. But I could take a look at it if you feel strongly about it...

hsoft commented:
> although test_selection() is still useful

I agree

> you could dump the entry_format constructor parameter and change the setEntryFormat class method to (say) setDMYEntry which takes a bool instead of a format

This is precisely what I meant in my previous comment.

> I'm a bit hesitant as the code for the more general case already has been written and I would argue that it's not that complicated

Code is written once and read many times. Re-writing to simplify is very often worth it. Individually, it's true that the difference in complexity is not *that* great, but this complexity adds up when someone new to the codebase tries to wrap her head around it.

Thanks again for this feature. I haven't found the time yet to do a proper review (try it for real), but I will do soon.

lenlo commented:
Simplified the code as per your request.

hsoft commented on core/tests/gui/date_widget_test.py:
```diff
@@ -431,6 +432,7 @@ def test_text(self):
 
 class TestCaseYYYYMMDDWithHyphen:
     def setup_method(self, method):
+        DateWidget.setDMYEntryOrder(False)
```
Is this needed because we call `setDMYEntryOrder(True)` in the previous test? If that's the case, this call doesn't go there, but rather in `TestCaseYYYYMMDDWithDot` teardown. Otherwise, we have tests depending on the order in which they're run and it's bad practice.

http://doc.pytest.org/en/latest/xunit_setup.html#method-and-function-level-setup-teardown

(as a side note, it's also one of the reasons why I'm lukewarm about class-level variables, but that's another subject...)

lenlo replied:
It's because this is the only other test case that doesn't use a d/m/y date order. For the rest of the test cases, it doesn't matter if DMY is set true or not.

What's the teardown method called? I can move it there instead. (I didn't know it existed.)

lenlo replied:
Never mind, I found it (`teardown_method`, d'oh).

hsoft commented on core/gui/date_widget.py:
```diff
@@ -25,72 +25,45 @@
     @property
     def _selected(self):
-        return self.__selected or self._selection_order[0]
+        return self.__selected or (DAY if self.__class__._dmyOrder else self._order[0])
```
Is there a specific reason why you use `self.__class__._dmyOrder`? Class-level variable can generally directly be read throuh the instance. Unless there's something else in the instance shadowing it, `self._dmyOrder` should work just fine.

lenlo replied:
OK, I didn't realize that.

hsoft commented:
I've tested it functionally on Qt and it generally works well. The only small problem I see is that the new option is `False` by default. It should be `True` to match previous behavior.

I'll be testing the Cocoa part soon and we'll be good to merge.

(sorry for the review delay, I've been busy)

hsoft commented:
Works fine on OS X too.

lenlo commented:
You're right, it should be `True` to match the previous behavior. However, is that really the right choice? I can understand the argument that those who have gotten used to the present behavior would rather not have it change underneath them, but what about all future moneyGuru users who likely will find an unexpected entry order confusing and annoying (like I did)? To me, the natural choice would be to have left-to-right entry as the default and let those who want a different entry order change it themselves. After all, that's how you enter dates in all other applications (and on paper, etc).

hsoft commented:
Default behavior is much more efficient, entry-wise. Most of the time, when you're entering many transactions, the only part of the date that changes is the day, so you're saving yourself 4 whole keystrokes. That ends up adding to a lot.

lenlo commented:
True, day-first entry can save you a couple of keystrokes, but only one or two as far as I can tell. To enter a different day of this month, you'd only need to press the right arrow once to get to the day subfield if your preferred date format is MM/DD/YYYY and twice if it's YYYY-MM-DD. That might be a good reason to have day-first entry available as an option, but as I see it, the more natural left-to-right order still ought to be the default -- just like automatic placement of a decimal mark is an option too, but the default is to enter it manually.

BTW, shouldn't it always be possible to enter the date separator character to move between the date components? It looks like you only can do that if you have manually entered a date component value today. E.g. with YYYY-MM-DD, just entering '-' doesn't do anything unless I first enter a year value, then '-' moves to the month. It seems to me that it should move to the month right away using the default value for the year. I don't think I changed that, but would you like me to roll that into this fix too?

hsoft commented:
What about going gradually? We start by introducing the new date entry mode, and then, in a subsequent release, make it the default if it's a popular demand. What do you think about that?

As for your suggestion, I think it's a good one. I can't think of a possible downside to it. So yes, if you want to work on this, I'll happily merge.

lenlo commented:
To be honest, you own this project so you can do what you like with it. But no, I don't think that's a good idea. I like simplicity and I like consistency between applications. I like to see dates formatted and entered the same way regardless of where I do it. It's fine to have features that break this as options, but they should be just that: options. The default should (IMHO) be the "normal" and expected way of operating. Only if the win in breaking a pattern clearly is significant would I think it worth pushing it onto the user from the start. In this case, it seems to save the occasional keystroke for (some) users, but not (again, IMHO) enough to warrant it being the default due to the consternation and confusion that comes with it.

But the choice is yours.

I'll look into making the other change we talked about.

lenlo commented:
I was about to say that I just removed the check for `_flush_buffer()` success before proceeding to go to the `_next()` subfield when entering the date format separator. It did the job and seemed harmless enough, but then I noticed the failed test. Could you tell me what this is about? Any reason why I shouldn't just replace "0 " with "12"?

```
    def test_type_sep(self):
        # There was a crash when a sep-caused _flush_buffer would be called with an invalid value
        self.w.type('/')
        eq_(self.w.text, '0 /06/2008') # don't do anything (stay on DAY)
```

hsoft commented:
It looks like the test comment is not descriptive enough. Before your change, the behavior was, once a partial date is entered, to only accept input that would complete that partial date. Back then, I felt that allowing a "revert to old value on invalid input" was too confusing behavior-wise.

But now that we allow separator character to be used to skip to the next field without entering a new value, the correct behavior of that test would indeed be to revert to its previous value and change its focus to the next field.

lenlo commented:
OK, thanks. I've updated the test accordingly.
